Sources/Testing/Polling/Polling.swift

```diff
@@ -0,0 +1,293 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+/// A type for managing polling
+@available(macOS 13, iOS 17, watchOS 9, tvOS 17, visionOS 1, *)
+struct Polling {
+  /// Run polling for a closure that evaluates to a boolean value.
+  ///
+  /// - Parameters:
+  ///   - behavior: The PollingBehavior to use.
+  ///   - timeout: How long to poll for until we time out.
+  ///   - closure: The closure to continuously evaluate.
+  ///   - expression: The expression, corresponding to `condition`, that is being
+  ///     evaluated (if available at compile time.)
+  ///   - comments: An array of comments describing the expectation. This array
+  ///     may be empty.
+  ///   - isRequired: Whether or not the expectation is required. The value of
+  ///     this argument does not affect whether or not an error is thrown on
+  ///     failure.
+  ///   - sourceLocation: The source location of the expectation.
+  static func run(
+    behavior: PollingBehavior,
+    timeout: Duration,
+    closure: @escaping @Sendable () async -> Bool,
+    expression: __Expression,
+    comments: [Comment],
+    isRequired: Bool,
+    sourceLocation: SourceLocation
+  ) async -> Result<Void, any Error> {
+    var expectation = Expectation(
+      evaluatedExpression: expression,
+      isPassing: true,
+      isRequired: isRequired,
+      sourceLocation: sourceLocation
+    )
+
+    let result = await poll(expression: closure, behavior: behavior, timeout: timeout)
+
+    let sourceContext = SourceContext(backtrace: nil, sourceLocation: sourceLocation)
+
+    switch result {
+    case .timedOut:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .timedOutWithoutRunning:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .finished:
+      return __checkValue(
+        true,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .failed:
+      return __checkValue(
+        false,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .cancelled:
+      Issue(
+        kind: .system,
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    }
```

Comment:
a) I'm going to deduplicate this analysis with the nearly-identical section in the other version of `run`. But time constraints & a desire to let others try out this code were why this was left in.
b) I'd really like to be able to provide additional explanations behind why each issue is recorded, but it wasn't clear what additional arguments to Issue (if any, maybe I need to set a property and I missed that?) I should use to provide that additional context.